### ISSUE TITLE
[RFC] SGLang IR A Functional Intermediate Representation for SGLang Operators

### ISSUE THREAD

**Issue by pre-master** [sgl-project/sglang#26889]
- **Status**: Draft
- **Target project**: SGLang
- **Authors**: TBD
- **Created**: 2026-06-01
- **Tracking issue / PR**: TBD
- **Related work**:
  - vLLM RFC: [vLLM IR: A Functional Intermediate Representation for vLLM](https://github.com/vllm-project/vllm/issues/32358)
  - vLLM implementation PR: [[vLLM IR] 1/N Implement IR skeleton and rms_norm op](https://github.com/vllm-project/vllm/pull/33825)
  - SGLang RFC: [Unified kernel fusion and torch.compile](https://github.com/sgl-project/sglang/issues/10118)
  - SGLang RFC: [Building towards a Hardware Abstraction Layer in SGLang](https://github.com/sgl-project/sglang/issues/26426)
  - SGLang issue: [Implement defunctionalization for sgl_kernel ops](https://github.com/sgl-project/sglang/issues/21550)
  - SGLang issue: [Replace memory-bound custom CUDA kernels](https://github.com/sgl-project/sglang/issues/21855)
  - SGLang issue: [Triton RMSNorm kernel underutilizes T4/A10G/A100 SMs](https://github.com/sgl-project/sglang/issues/24353)

> **Attribution**: This RFC is explicitly inspired by and structurally references vLLM PR [#33825](https://github.com/vllm-project/vllm/pull/33825), which implements the initial vLLM IR skeleton and an `rms_norm` IR operator. The proposed SGLang design adapts the same high-level ideas鈥攆unctional IR operators, native semantic definitions, implementation registration, priority-based dispatch, and lowering鈥攊nto SGLang's serving, kernel, and compilation architecture. This RFC does not propose copying vLLM code verbatim; it proposes adopting the architectural pattern where it fits SGLang.

---

## 1. Summary

This RFC proposes introducing **SGLang IR**, a functional intermediate representation for selected SGLang operators such as `rms_norm`, `fused_add_rms_norm`, `rope`, `activation`, `quant`, `kv_cache_store`, and other memory-bound or fusion-sensitive operators.

The key idea is to separate:

1. **Operator semantics**: what the operator means mathematically.
2. **Operator implementations**: how the operator is implemented on CUDA, ROCm, XPU, Triton, CUTLASS/CuTe, FlashInfer, `sgl-kernel`, or native PyTorch.
3. **Dispatch policy**: which implementation should be selected for a given platform and argument set.
4. **Compilation representation**: what `torch.compile` / FX / Inductor should see before lowering.
5. **Lowering**: when and how the high-level IR operator is replaced by a concrete implementation.

Instead of exposing many backend-specific custom operators directly to model code and compiler passes, SGLang IR would expose stable high-level functional operators. Compiler passes can then match these IR operators directly, and a later lowering pass can choose the concrete kernel implementation.

---

## 2. Motivation

SGLang already contains multiple kernel paths and optimization efforts around custom kernels, Triton kernels, CUDA graph capture, `torch.compile`, fusion, and hardware abstraction. These efforts expose a recurring architectural problem: high-level operator semantics are often coupled with backend-specific implementations.

This makes the system harder to optimize, test, and extend.

#### 2.1.1 Compiler passes see implementation details too early

When model code calls concrete custom kernels directly, `torch.compile` and FX passes may see backend-specific operators or decomposed PyTorch graphs instead of stable semantic operators.

For example, an RMSNorm operation may appear as:

- a native PyTorch decomposition,
- an `sgl-kernel` custom op,
- a Triton implementation,
- a FlashInfer-backed implementation,
- a fused variant,
- or an implementation hidden behind Python dispatch logic.

This makes fusion and pattern-matching fragile. A compiler pass that wants to optimize `RMSNorm + residual add` should not need to know every backend-specific implementation detail.

#### 2.1.2 Operator semantics and kernel implementations are coupled

SGLang supports and experiments with multiple kernel providers. However, when operator calls are wired directly to specific providers, it becomes difficult to define a single source of truth for correctness.

For example, a single operator such as RMSNorm has subtle semantic details:

- whether weight is optional,
- whether computation is performed in fp32,
- whether output dtype follows input dtype,
- whether residual input is mutated,
- whether the operation is in-place, out-of-place, or maybe-in-place,
- whether shape constraints differ across backends.

Without a semantic IR layer, each implementation can silently encode slightly different assumptions.

#### 2.1.3 Backend selection is scattered

SGLang increasingly needs to select among different backend implementations depending on:

- GPU architecture,
- CUDA / ROCm / XPU availability,
- tensor dtype,
- tensor shape,
- sequence length,
- capture mode,
- compilation mode,
- whether an implementation is available in the current build,
- and user preferences.

A unified IR-level priority mechanism would make this selection explicit, configurable, testable, and easier to reason about.

#### 2.1.4 Existing SGLang issues point to the same need

Recent SGLang discussions already point toward similar requirements:

- kernel fusion and `torch.compile` require operators to be represented in a compiler-friendly way;
- a hardware abstraction layer requires platform-dependent modules to be wrapped behind a dispatcher;
- defunctionalization is needed to avoid unnecessary tensor copies around custom in-place kernels;
- memory-bound decode-path kernels such as RMSNorm, RoPE, q/k normalization, and KV-cache stores remain important optimization targets.

SGLang IR would provide a common abstraction for these efforts.

---

## 3. Goals

SGLang IR should provide:

1. **Stable operator semantics**  
   Each IR operator has a native semantic implementation that defines correctness.

2. **Functional representation**  
   IR operators are represented as functional operators by default, so compiler passes can reason about them safely.

3. **Implementation registration**  
   Backend providers can register concrete implementations for an IR operator.

4. **Priority-based dispatch**  
   SGLang can choose the best available implementation through platform defaults and user overrides.

5. **Compiler-friendly tracing**  
   `torch.compile` should see high-level SGLang IR operators before lowering, not a mixture of unrelated backend-specific kernels.

6. **Late lowering**  
   IR operators should be lowered to concrete implementations after high-level fusion and transformation passes.

7. **Incremental migration**  
   The system should allow one operator at a time to migrate into SGLang IR without requiring a full rewrite.

8. **Out-of-tree extensibility**  
   External kernel packages should eventually be able to register implementations without modifying SGLang core.

---

## 4. Non-goals

This RFC does not propose:

1. Replacing all SGLang kernels immediately.
2. Rewriting the entire model execution stack.
3. Replacing `torch.compile`, FX, Inductor, Triton, or `sgl-kernel`.
4. Defining a general-purpose MLIR dialect.
5. Requiring every operator to become an IR operator.
6. Removing backend-specific optimized kernels.
7. Guaranteeing that the first implementation improves latency.

The first milestone should focus on infrastructure correctness and low-risk migration, using `rms_norm` as the initial operator.

---

## 5. Design overview

The proposed architecture follows this pipeline:

```mermaid
flowchart TD
    A[Model code calls sg_ir.ops.rms_norm] --> B[Dynamo / FX tracing]
    B --> C[FX graph contains torch.ops.sgl_ir.rms_norm.default]
    C --> D[High-level SGLang compiler passes]
    D --> E[IR fusion / transformation]
    E --> F[SGLang IR lowering pass]
    F --> G{Priority-based implementation selection}
    G --> H1[sgl_kernel implementation]
    G --> H2[Triton implementation]
    G --> H3[FlashInfer / vendor implementation]
    G --> H4[Native PyTorch fallback]
    H1 --> I[Inductor / runtime execution]
    H2 --> I
    H3 --> I
    H4 --> I
```

The essential design is:

- `register_op` defines an IR operator and its native semantics.
- `register_impl` registers concrete implementations.
- `supports_args` determines whether a concrete implementation can handle the current arguments.
- `ir_op_priority` determines implementation preference.
- eager mode dispatch selects an implementation at runtime.
- compiled mode keeps the IR op visible until a lowering pass replaces it with the selected implementation.

---

### 6.1 Defining an IR operator

An IR operator is registered with a native semantic implementation.

```python
# python/sglang/ir/ops/layernorm.py

import torch
from torch import Tensor

from sglang.ir import register_op


@register_op("rms_norm")
def rms_norm_native(
    x: Tensor,
    weight: Tensor | None,
    eps: float,
) -> Tensor:
    """Native semantic definition of RMSNorm.

    This implementation is the semantic source of truth.
    Backend implementations must match this behavior.
    """
    variance = x.to(torch.float32).pow(2).mean(dim=-1, keepdim=True)
    out = x * torch.rsqrt(variance + eps)

    if weight is not None:
        out = out * weight

    return out.to(dtype=x.dtype)
```

The native implementation has three roles:

1. **Correctness reference** for tests.
2. **Fallback implementation** when no optimized implementation is available.
3. **Semantic definition** for all backend providers.

### 6.2 Calling an IR operator from model code

Model code should call the IR operator rather than a provider-specific kernel.

```python
from sglang import ir

hidden_states = ir.ops.rms_norm(hidden_states, weight, eps)
```

In eager mode, this call dispatches to the selected implementation.

In compiled mode, this call should trace as a high-level IR op:

```python
torch.ops.sgl_ir.rms_norm.default(hidden_states, weight, eps)
```

### 6.3 Registering an implementation

Concrete implementations are registered against the IR operator.

```python
# python/sglang/kernels/rms_norm.py

from sglang import ir
from sglang.kernels import sgl_kernel_rms_norm


def supports_sgl_kernel_rms_norm(x, weight, eps):
    return (
        x.is_cuda
        and x.dtype in (torch.float16, torch.bfloat16)
        and x.shape[-1] <= 8192
    )


@ir.ops.rms_norm.register_impl(
    provider="sgl_kernel",
    supports_args=supports_sgl_kernel_rms_norm,
    inplace=False,
)
def rms_norm_sgl_kernel(x, weight, eps):
    return sgl_kernel_rms_norm(x, weight, eps)
```

### 6.4 Registering a Triton implementation

```python
@ir.ops.rms_norm.register_impl(
    provider="triton",
    supports_args=lambda x, weight, eps: (
        x.is_cuda and x.dtype in (torch.float16, torch.bfloat16)
    ),
    inplace=False,
)
def rms_norm_triton(x, weight, eps):
    return triton_rms_norm(x, weight, eps)
```

### 6.5 Registering a native fallback

The native implementation is automatically available as `native`.

```python
# Conceptually available by default:
provider = "native"
implementation = rms_norm_native
```

The native implementation should be the last fallback in most priority lists.

---

## 7. Priority-based dispatch

SGLang should provide an IR op priority configuration.

### 7.1 CLI

Proposed CLI:

```bash
python3 -m sglang.launch_server \
  --model-path meta-llama/Llama-3.1-8B-Instruct \
  --ir-op-priority.rms_norm=sgl_kernel,triton,native
```

Multiple operators can be configured independently:

```bash
python3 -m sglang.launch_server \
  --model-path meta-llama/Llama-3.1-8B-Instruct \
  --ir-op-priority.rms_norm=sgl_kernel,triton,native \
  --ir-op-priority.rope=triton,sgl_kernel,native \
  --ir-op-priority.activation=triton,native
```

### 7.2 Python config

```python
server_args.ir_op_priority = {
    "rms_norm": ["sgl_kernel", "triton", "native"],
    "rope": ["triton", "sgl_kernel", "native"],
}
```

### 7.3 Platform defaults

Platforms can define default priorities.

```python
CUDA_DEFAULT_IR_OP_PRIORITY = {
    "rms_norm": ["sgl_kernel", "triton", "native"],
    "fused_add_rms_norm": ["sgl_kernel", "triton", "native"],
    "rope": ["sgl_kernel", "triton", "native"],
}

ROCM_DEFAULT_IR_OP_PRIORITY = {
    "rms_norm": ["triton", "native"],
    "fused_add_rms_norm": ["triton", "native"],
}

CPU_DEFAULT_IR_OP_PRIORITY = {
    "rms_norm": ["native"],
}
```

User-provided priority should be prepended to platform defaults, then deduplicated.

Example:

```text
User priority:      rms_norm=triton
Platform default:   rms_norm=sgl_kernel,triton,native
Effective priority: rms_norm=triton,sgl_kernel,native
```

### 7.4 Dispatch algorithm

```python
def dispatch(ir_op, *args, **kwargs):
    priority = ir_op.get_effective_priority()

    for provider in priority:
        impl = ir_op.get_impl(provider)

        if impl is None:
            continue

        if not impl.is_supported_platform():
            continue

        if not impl.supports_args(*args, **kwargs):
            continue

        return impl(*args, **kwargs)

    raise RuntimeError(
        f"No valid implementation found for IR op {ir_op.name}. "
        f"Priority={priority}."
    )
```

`supports_args` should only check stable properties such as:

- device type,
- dtype,
- rank,
- hidden size,
- alignment,
- optional argument presence,
- compile-mode support,
- capture-mode support if statically known.

It should not branch on dynamic batch size, request count, or values that would create unstable compile guards.

---

## 8. Functionalization and maybe-in-place operators

Some SGLang kernels mutate outputs or reuse input buffers for performance. However, compiler passes are easier to reason about when operators are functional.

This RFC proposes two overload modes:

1. `default`: functional, no mutation visible to the graph.
2. `maybe_inplace`: allows buffer donation or reuse when safe.

### 8.1 Example

```python
hidden_states = ir.ops.rms_norm.default(hidden_states, weight, eps)

hidden_states = ir.ops.rms_norm.maybe_inplace(
    hidden_states,
    weight,
    eps,
    donated_inputs=(0,),
)
```

### 8.2 Compiled-mode behavior

In compiled mode:

1. `maybe_inplace` is first represented as an IR op.
2. A SGLang functionalization pass rewrites it to `default`.
3. The pass records donation metadata.
4. Later clone-elimination or defunctionalization can restore memory reuse where safe.
5. The final lowering pass selects a concrete implementation.

```mermaid
flowchart LR
    A[maybe_inplace IR op] --> B[Functionalization]
    B --> C[default IR op + donation metadata]
    C --> D[Optimization passes]
    D --> E[Clone elimination / defunctionalization]
    E --> F[Lowering to concrete kernel]
```

This design aligns with SGLang's existing need to avoid unnecessary copies around custom in-place kernels while preserving compiler correctness.

---

## 9. Lowering pass

SGLang should add an IR lowering pass that runs after high-level IR fusion and transformation passes.

### 9.1 Lowering responsibilities

The lowering pass should:

1. Find `torch.ops.sgl_ir.*` nodes in the FX graph.
2. Resolve the effective priority list for each IR op.
3. Select the first compatible implementation.
4. Replace the IR op node with the selected implementation graph or custom op call.
5. Preserve metadata needed by Inductor, CUDA graph capture, and runtime execution.
6. Include lowering-affecting configuration in the compile cache key.

### 9.2 Example before lowering

```python
def forward(hidden_states, weight):
    return torch.ops.sgl_ir.rms_norm.default(hidden_states, weight, 1e-6)
```

### 9.3 Example after lowering

```python
def forward(hidden_states, weight):
    return torch.ops.sgl_kernel.rms_norm(hidden_states, weight, 1e-6)
```

or:

```python
def forward(hidden_states, weight):
    return triton_rms_norm(hidden_states, weight, 1e-6)
```

or, if no optimized implementation supports the arguments:

```python
def forward(hidden_states, weight):
    return rms_norm_native(hidden_states, weight, 1e-6)
```

### 9.4 Compile cache key

The lowering output depends on:

- selected provider,
- provider priority list,
- implementation source,
- implementation version,
- platform defaults,
- relevant feature flags,
- and possibly CUDA / ROCm / XPU capability.

Therefore, the compile cache key should include an IR lowering fingerprint.

```python
def ir_lowering_fingerprint():
    return hash((
        effective_ir_op_priority,
        registered_impl_uuids,
        platform_name,
        platform_capability,
        relevant_feature_flags,
    ))
```

This prevents reusing a compiled graph that was generated under a different IR implementation policy.

---

## 10. Proposed package structure

```text
python/sglang/
  ir/
    __init__.py
    op.py
    registry.py
    config.py
    lowering_pass.py
    functionalization.py
    clone_elimination.py
    ops/
      __init__.py
      layernorm.py
      activation.py
      rope.py
      quant.py
      kv_cache.py
  kernels/
    ir_impls/
      __init__.py
      rms_norm_sgl_kernel.py
      rms_norm_triton.py
      rope_sgl_kernel.py
      activation_triton.py
```

### 10.1 Core modules

| Module | Responsibility |
|---|---|
| `sglang.ir.op` | Defines `IrOp`, `IrOpImpl`, overloads, dispatch helpers |
| `sglang.ir.registry` | Global registry for IR ops and implementations |
| `sglang.ir.config` | CLI and Python config for IR priority |
| `sglang.ir.lowering_pass` | FX graph lowering from IR ops to concrete implementations |
| `sglang.ir.functionalization` | Rewrites maybe-in-place IR ops into functional form |
| `sglang.ir.clone_elimination` | Removes redundant clones introduced by functionalization |
| `sglang.ir.ops.*` | Native semantic definitions of IR operators |
| `sglang.kernels.ir_impls.*` | Provider-specific registered implementations |

---
